### DISCUSSION TITLE
Error loading Python lib '/tmp/_MEIruWkD7/libpython3.9.so.1.0': dlopen: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.35' not found (required by /tmp/_MEIruWkD7/libpython3.9.so.1.0

### DISCUSSION THREAD

**Discussion by CloudWise-Lukemiao** [pyinstaller/pyinstaller#7541] [Help]
[1969296] Error loading Python lib '/tmp/_MEIruWkD7/libpython3.9.so.1.0': dlopen: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.35' not found (required by /tmp/_MEIruWkD7/libpython3.9.so.1.0)

I want to use Github Action to publish my Python program, which uses pyinstaller for packaging. But I found that it cannot run on other Linux machines. What should I do to make it run properly. Here is the yml file content for my github action

name: Build and Package

on:

push:

branches:

- main

jobs:

build:

runs-on: ubuntu-latest

steps:

- name: Checkout code

uses: actions/checkout

- name: Set up Python

uses: actions/setup-python

with:

python-version: '3.9'

- name: Install dependencies

run: |

python -m pip install --upgrade pip

pip3 install -r requirements.txt

- name: Build package

run: |

pyinstaller iotd.py --hidden-import tendo --hidden-import yaml --hidden-import paramiko --hidden-import prettytable --hidden-import ruamel.yaml --hidden-import rich --hidden-import rich.progress --hidden-import logzero --hidden-import tqdm --add-data 'cluster/*:cluster' --add-data 'tools/*:tools' --add-data 'exception/*:exception' --onefile --runtime-tmpdir /tmp

- name: Copy files to iotd/sbin and iotd/config

run: |

mkdir -p iotd/sbin

mkdir -p iotd/config

cp -r dist/* iotd/sbin/

cp -r config/* iotd/

- name: Upload artifacts

uses: actions/upload-artifact

with:

name: iotd

path: ./ iotd

**Comment by rokm** [comment]
You cannot run your frozen application on linux distributions that have older version of `glibc` than your build system (because the python shared library is linked against that particular `glibc` version, and expects symbols of at least that version - as you can see in your error message). So you need to build on linux distribution that has `glibc` as old as your target distributions.

So you probably shouldn't be building on `ubuntu-latest` (= `ubuntu-22.04`); `ubuntu-20.04` should cover a wider range of distributions, although the application still won't run on distributions that are older than that.

**Reply by CloudWise-Lukemiao** [reply]
In the future, I want to use pyinstaller to package programs and run them on Centos7. What should I do?

**Reply by rokm** [reply]
Package them under CentOS 7, then - and set up your build system in a docker container, if you want to run it on GitHub Actions' ubuntu runner.

**Comment by sivakumar41** [comment]
> [1969296] Error loading Python lib '/tmp/_MEIruWkD7/libpython3.9.so.1.0': dlopen: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.35' not found (required by /tmp/_MEIruWkD7/libpython3.9.so.1.0)
> 
> I want to use Github Action to publish my Python program, which uses pyinstaller for packaging. But I found that it cannot run on other Linux machines. What should I do to make it run properly. Here is the yml file content for my github action
> 
> name: Build and Package
> 
> on:
> 
> push:
> 
> branches:
> 
> * main
> 
> jobs:
> 
> build:
> 
> runs-on: ubuntu-latest
> 
> steps:
> 
> * name: Checkout code
> 
> uses: actions/checkout
> 
> * name: Set up Python
> 
> uses: actions/setup-python
> 
> with:
> 
> python-version: '3.9'
> 
> * name: Install dependencies
> 
> run: |
> 
> python -m pip install --upgrade pip
> 
> pip3 install -r requirements.txt
> 
> * name: Build package
> 
> run: |
> 
> pyinstaller iotd.py --hidden-import tendo --hidden-import yaml --hidden-import paramiko --hidden-import prettytable --hidden-import ruamel.yaml --hidden-import rich --hidden-import rich.progress --hidden-import logzero --hidden-import tqdm --add-data 'cluster/_:cluster' --add-data 'tools/_:tools' --add-data 'exception/*:exception' --onefile --runtime-tmpdir /tmp
> 
> * name: Copy files to iotd/sbin and iotd/config
> 
> run: |
> 
> mkdir -p iotd/sbin
> 
> mkdir -p iotd/config
> 
> cp -r dist/* iotd/sbin/
> 
> cp -r config/* iotd/
> 
> * name: Upload artifacts
> 
> uses: actions/upload-artifact
> 
> with:
> 
> name: iotd
> 
> path: ./ iotd

@CloudWise-Lukemiao  did you solved this problem?

**Comment by somegenericdev** [comment]
@rokm I'm wondering, is there no way GLIBC could be statically linked with the interpreter and packaged in the executable?

**Reply by rokm** [reply]
It depends on what you are trying to package.

For simple programs that do not need to load anything from the system, you can use `staticx` in combination with PyInstaller's `onefile` build.

But I don't think that will work for anything more complex, for example if you use GUI framework that needs to load some of system-provided shared libraries (e.g., graphics driver stack).
